crates/primitives-traits/src/signed_tx.rs

```diff
@@ -0,0 +1,138 @@
+
+use crate::Signature;
+
+/// A signed transaction.
+pub trait SignedTransaction:
+    fmt::Debug
+    + Clone
+    + PartialEq
+    + Eq
+    + Hash
+    + Send
+    + Sync
+    + serde::Serialize
+    + for<'a> serde::Deserialize<'a>
+    + alloy_rlp::Encodable
+    + alloy_rlp::Decodable
+    + Encodable2718
+    + Decodable2718
+{
+    /// Transaction type that is signed.
+    type Transaction: SignableTransaction<Self::Signature>;
+
+    /// Signature type that results from signing transaction.
+    type Signature: Signature;
+
+    /// Returns reference to transaction hash.
+    fn tx_hash(&self) -> &TxHash;
+
+    /// Returns reference to transaction.
+    fn transaction(&self) -> &Self::Transaction;
+
+    /// Returns reference to signature.
+    fn signature(&self) -> &Self::Signature;
+
+    /// Recover signer from signature and hash.
+    ///
+    /// Returns `None` if the transaction's signature is invalid following [EIP-2](https://eips.ethereum.org/EIPS/eip-2), see also `reth_primitives::transaction::recover_signer`.
+    ///
+    /// Note:
+    ///
+    /// This can fail for some early ethereum mainnet transactions pre EIP-2, use
+    /// [`Self::recover_signer_unchecked`] if you want to recover the signer without ensuring that
+    /// the signature has a low `s` value.
+    fn recover_signer(&self) -> Option<Address>;
+
+    /// Recover signer from signature and hash _without ensuring that the signature has a low `s`
+    /// value_.
+    ///
+    /// Returns `None` if the transaction's signature is invalid, see also
+    /// `reth_primitives::transaction::recover_signer_unchecked`.
+    fn recover_signer_unchecked(&self) -> Option<Address>;
+
+    /// Output the length of the `encode_inner(out`, true). Note to assume that `with_header` is
+    /// only `true`.
+    fn payload_len_inner(&self) -> usize;
+
+    /// Returns the length without an RLP header - this is used for eth/68 sizes.
+    fn length_without_header(&self) -> usize;
+
+    /// Decodes legacy transaction from the data buffer.
+    ///
+    /// This should be used _only_ be used in general transaction decoding methods, which have
+    /// already ensured that the input is a legacy transaction with the following format:
+    /// `rlp(legacy_tx)`
+    ///
+    /// Legacy transactions are encoded as lists, so the input should start with a RLP list header.
+    ///
+    /// This expects `rlp(legacy_tx)`
+    // TODO: make buf advancement semantics consistent with `decode_enveloped_typed_transaction`,
+    // so decoding methods do not need to manually advance the buffer
+    fn decode_rlp_legacy_transaction(data: &mut &[u8]) -> alloy_rlp::Result<Self>;
+
+    /// Decodes legacy transaction from the data buffer into a tuple.
+    ///
+    /// This expects `rlp(legacy_tx)`
+    ///
+    /// Refer to the docs for [`Self::decode_rlp_legacy_transaction`] for details on the exact
+    /// format expected.
+    fn decode_rlp_legacy_transaction_tuple(
```

Comment:
I consider all of this out of scope for now, doesn't seem reasonable to include this here